admin-dev/themes/new-theme/js/pages/order/view/order-view-page.js

```diff
@@ -337,4 +308,68 @@ export default class OrderViewPage {
   getActivePage() {
     return $(OrderViewPageMap.productsTablePagination).find('.active span').get(0);
   }
+
+  refreshProductsList(orderId) {
+    const $loadingSpinner = $(OrderViewPageMap.productsPanel).find('.spinner-order-products-container#orderProductsLoading');
+    $loadingSpinner.show();
+
+    const $tablePagination = $(OrderViewPageMap.productsTablePagination);
+    const numRowsPerPage = $tablePagination.data('numPerPage');
+    const initialNumProducts = $(OrderViewPageMap.productsTableRows).length;
+
+    $.ajax(this.router.generate('admin_orders_get_products', {orderId}))
+        .done((response) => {
+          // Delete previous product lines
+          $(OrderViewPageMap.productsTable).find(OrderViewPageMap.productsTableRows).remove();
+
+          $(OrderViewPageMap.productsTable + ' tbody').prepend(response);
+
+          $loadingSpinner.hide();
+
+          const newNumProducts = $(OrderViewPageMap.productsTableRows).length;
+          const initialPagesNum = Math.max(Math.ceil(initialNumProducts / numRowsPerPage), 1);
+          const newPagesNum = Math.ceil(newNumProducts / numRowsPerPage);
+
+          this.orderProductRenderer.updateNumProducts(newNumProducts);
+
+          // Update pagination
+          if (newPagesNum > initialPagesNum) {
+            this.orderProductRenderer.paginationAddPage(newPagesNum);
+          }
+
+          // Display alert
+          const $alertBlock = $('#js-order-view-success-block');
+          if (initialNumProducts > newNumProducts) {
+            $alertBlock.find('.alert-text').html(
+                window.translate_javascripts['Items successfully removed']
+                    .replace('[1]', (initialNumProducts-newNumProducts))
+                    .replace('[2]', (initialNumProducts-newNumProducts) > 1 ? 's' : '')
+            );
+            this._fadeAlert($alertBlock);
+          }
+          else if (initialNumProducts < newNumProducts) {
+            $alertBlock.find('.alert-text').html(
+                window.translate_javascripts['Items successfully added']
+                    .replace('[1]', (newNumProducts-initialNumProducts))
+                    .replace('[2]', (newNumProducts-initialNumProducts) > 1 ? 's' : '')
+            );
+            this._fadeAlert($alertBlock);
+          }
+
+          // Bind actions on products
+          this.listenForProductDelete();
```

Comment:
As @eternoendless said you should modify the listeners so that they are registered on the parent but with sub selectors This will allow you to only listen once and then no need to re-register the listeners With jquery it's quite easy, it's called delegated selectors, instead of doing:

```js
$(OrderViewPageMap.productDeleteBtn)
      .off('click')
      .on('click', (event) => this.orderProductManager.handleDeleteProductEvent(event));
```

you can do

```js
$(OrderViewPageMap.productsTable).on(
    'click',
    OrderViewPageMap.productDeleteBtn,
    (event) => this.orderProductManager.handleDeleteProductEvent(event)
);
```

The idea is that `OrderViewPageMap.productsTable` which is the element that will not be removed is the one that listens the `click` event even when it bubbles from a child, and it only triggers the handler when the original target of the even is `OrderViewPageMap.productDeleteBtn`

So you register only **once** and then no matter how many elements you add or remove the click is already registered